### PR TITLE
Job deduplication and priority

### PR DESCRIPTION
**Pull request by panda-sheep**
As title

### CODE REVIEW

File: src/meta/processors/jobMan/JobManager.h [hunk 1]
```diff
@@ -79,7 +101,15 @@ class JobManager : public nebula::cpp::NonCopyable, public nebula::cpp::NonMovab
...
 private:
-    std::unique_ptr<folly::UMPSCQueue<int32_t, true>>  queue_;
+    // Todo(pandasheep)
+    // When folly is upgraded, PriorityUMPSCQueueSet can be used
+    // Use two queues to simulate priority queue, Divide by job cmd
+    std::unique_ptr<folly::UMPSCQueue<JobID, true>>    lowPriorityQueue_;
+    std::unique_ptr<folly::UMPSCQueue<JobID, true>>    highPriorityQueue_;
+
+    // The job in running or queue
+    folly::ConcurrentHashMap<JobID, JobDescription>    jobMap_;
```
Comment (liuyu85cn) [inline src/meta/processors/jobMan/JobManager.h:111]: may be we can called this 'inFlightJobs' ?

Comment (panda-sheep) [reply src/meta/processors/jobMan/JobManager.h:111]: Good point, 👍

**Review by critical27 — APPROVED** [review]

**Review by liuyu85cn — APPROVED** [review]
LGTM

File: src/meta/processors/jobMan/JobManager.cpp [hunk 2]
```diff
@@ -170,6 +181,27 @@ cpp2::ErrorCode JobManager::addJob(const JobDescription& jobDesc, AdminClient* c
...
+    if (highPriorityQueue_->try_dequeue(jobId)) {
+        return true;
+    } else if (lowPriorityQueue_->try_dequeue(jobId)) {
+        return true;
+    }
+    return false;
+}
+
+void JobManager::enqueue(const JobID& jobId, const cpp2::AdminCmd& cmd) {
+    if (cmd == cpp2::AdminCmd::STATIS) {
```
Comment (darionyaphet) [inline src/meta/processors/jobMan/JobManager.cpp:198]: seems statis job will take higher priority ?

File: src/meta/processors/jobMan/JobManager.cpp [hunk 3]
```diff
@@ -233,6 +265,22 @@ void JobManager::removeExpiredJobs(const std::vector<std::string>& expiredJobsAn
     baton.wait();
 }
 
+bool JobManager::checkJobExist(const cpp2::AdminCmd& cmd,
+                               const std::vector<std::string>& paras,
+                               JobID& iJob) {
+    JobDescription jobDesc(0, cmd, paras);
+    auto it = inFlightJobs_.begin();
+    while (it != inFlightJobs_.end()) {
+        if (it->second == jobDesc) {
```
Comment (darionyaphet) [inline src/meta/processors/jobMan/JobManager.cpp:274]: why don't compare job type and parameter directly ?

File: src/meta/processors/jobMan/JobManager.cpp (hunk 2 above)
Comment (panda-sheep) [reply src/meta/processors/jobMan/JobManager.cpp:198]: Yes, according to AdminCmd to determine whether it is high priority or low priority. 
Currently AdminCmd::STATIS is a high priority

File: src/meta/processors/jobMan/JobManager.cpp (hunk 3 above)
Comment (panda-sheep) [reply src/meta/processors/jobMan/JobManager.cpp:274]: Here is the condition for judging that JobDescription is equal:
job type, parameter and status

**Review by darionyaphet — APPROVED** [review]

**Pull request merged**